File: demo/ffi/ydraw/python/yimage/gallery.py
#!/usr/bin/env python3
"""ydraw client interface target sketch — yimage complex drawables.

NOT RUNNABLE YET. Conceptual: an Image packs one yimage complex record
(bounds + image_w/image_h + the pixel buffer feeding the atlas texture;
decode from PNG happens client-side, exactly where ycat does it today).
Assets: demo/assets/yimage/.
"""
from pathlib import Path

from yetty.ydraw import DrawableList, Image, Text

ASSETS = Path(__file__).resolve().parents[4] / "assets" / "yimage"


def show(*drawables):
    dlist = DrawableList()
    for drawable in drawables:
        dlist.add(drawable)
    dlist.dcs_emit()
    dlist.destroy()


# One image, displayed at its natural size.
print('rose')
show(Image(ASSETS / "rose.png"))

# Scaled display bounds: the record's bounds are independent of the
# pixel dimensions.
print('hero, scaled to 480px wide')
show(Image(ASSETS / "hero.png", width=480, height=270))

# The FFI advantage over the CLI: an image is just one record among
# others — caption it in the same envelope, in the same coordinate
# space.
print('wordmark with caption')
show(Image(ASSETS / "wordmark.png", x=0, y=0, width=320, height=96),
     Text("terminal unchained", x=0, y=112, font_size=18,
          color="#9FA7A8"))

# A row of thumbnails — layout is plain arithmetic, like everywhere else.
print('thumbnail strip')
strip = DrawableList()
for index, name in enumerate(["gradient.png", "rose.png", "hero.png",
                              "wordmark.png"]):
    strip.add(Image(ASSETS / name, x=index * 170, y=0,
                    width=160, height=100))
strip.dcs_emit()
strip.destroy()

File: demo/ffi/ydraw/go/yimage/gallery.go
// ydraw client interface target sketch — yimage complex drawables (Go).
// NOT RUNNABLE YET — see python/yimage/gallery.py. Asset paths are
// repo-relative. Assets: demo/assets/yimage/.
package main

import (
	"fmt"

	"github.com/zokrezyl/yetty/bindings/go/ydraw"
)

const assets = "demo/assets/yimage/"

func show(drawables ...ydraw.Drawable) {
	dlist := ydraw.NewDrawableList()
	for _, drawable := range drawables {
		dlist.Add(drawable)
	}
	dlist.DcsEmit()
	dlist.Destroy()
}

func main() {
	fmt.Println("rose")
	show(ydraw.Image{Path: assets + "rose.png"})

	fmt.Println("hero, scaled to 480px wide")
	show(ydraw.Image{Path: assets + "hero.png", Width: 480, Height: 270})

	// One record among others: caption in the same envelope.
	fmt.Println("wordmark with caption")
	show(ydraw.Image{Path: assets + "wordmark.png", X: 0, Y: 0, Width: 320, Height: 96},
		ydraw.Text{Body: "terminal unchained", X: 0, Y: 112, FontSize: 18,
			Color: "#9FA7A8"})

	fmt.Println("thumbnail strip")
	strip := ydraw.NewDrawableList()
	names := []string{"gradient.png", "rose.png", "hero.png", "wordmark.png"}
	for index, name := range names {
		strip.Add(ydraw.Image{Path: assets + name, X: float64(index) * 170,
			Y: 0, Width: 160, Height: 100})
	}
	strip.DcsEmit()
	strip.Destroy()
}
